File: companies/Informatica/200.number-of-islands.py
"""
Approach: Flood-fill each unvisited land component.
Data structure: an explicit stack performs DFS and mutates visited land to water to avoid a separate visited set.
Interview logic: every time we find a '1', it starts one island. DFS marks all land connected to it, so no cell from that island is counted again.
Complexity: O(mn) time, O(mn) worst-case stack space.
Tests and edge cases: empty grid returns 0; all water returns 0; diagonal adjacency does not connect islands.
"""
from __future__ import annotations
from typing import List

# @lc code=start
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        if not grid or not grid[0]:
            return 0
        rows, cols = len(grid), len(grid[0])
        islands = 0
        for r in range(rows):
            for c in range(cols):
                if grid[r][c] != '1':
                    continue
                islands += 1
                grid[r][c] = '0'
                stack = [(r, c)]
                while stack:
                    x, y = stack.pop()
                    for dx, dy in ((1, 0), (-1, 0), (0, 1), (0, -1)):
                        nx, ny = x + dx, y + dy
                        if 0 <= nx < rows and 0 <= ny < cols and grid[nx][ny] == '1':
                            grid[nx][ny] = '0'
                            stack.append((nx, ny))
        return islands
# @lc code=end

File: companies/Sprinklr/200.number-of-islands.go
package leetcode

// NumIslands200 scans for land, increments island count, and sinks that whole
// connected component with an explicit DFS stack. Mutating visited land to '0'
// keeps auxiliary visited space unnecessary.
//
// Time: O(m*n). Space: O(m*n) worst-case stack.
func NumIslands200(grid [][]byte) int {
	if len(grid) == 0 || len(grid[0]) == 0 {
		return 0
	}
	rows, cols := len(grid), len(grid[0])
	islands := 0
	dirs := [][2]int{{1, 0}, {-1, 0}, {0, 1}, {0, -1}}
	type cell struct{ r, c int }
	for r := 0; r < rows; r++ {
		for c := 0; c < cols; c++ {
			if grid[r][c] != '1' {
				continue
			}
			islands++
			grid[r][c] = '0'
			st := []cell{{r, c}}
			for len(st) > 0 {
				cur := st[len(st)-1]
				st = st[:len(st)-1]
				for _, d := range dirs {
					nr, nc := cur.r+d[0], cur.c+d[1]
					if nr >= 0 && nr < rows && nc >= 0 && nc < cols && grid[nr][nc] == '1' {
						grid[nr][nc] = '0'
						st = append(st, cell{nr, nc})
					}
				}
			}
		}
	}
	return islands
}
